### PR TITLE
added fee in elastic search transactions

### PR DESCRIPTION
**Pull request by miiu96**
Added fee in transaction that is indexed in elasticsearch database.
Added `gasPriceModifer` in the api route `/network/config`

### CODE REVIEW

File: process/economics/economicsData.go [hunk 1]
```diff
@@ -389,6 +389,44 @@ func (ed *economicsData) ComputeGasLimit(tx process.TransactionWithFeeHandler) u
...
+
+	computeFeeForProcessing := ed.ComputeFeeForProcessing(tx, gasUsed-moveBalanceGasLimit)
+	txFee := big.NewInt(0).Add(moveBalanceFee, computeFeeForProcessing)
+
+	return txFee
+}
+
+func (ed *economicsData) computeGasRefund(refundValue *big.Int, gasPrice uint64) uint64 {
+	gasPriceBig := big.NewInt(0).SetUint64(gasPrice)
+
```
Comment (sasurobert) [inline process/economics/economicsData.go:422]: delete empty line

File: process/economics/economicsData.go [hunk 2]
```diff
@@ -389,6 +389,44 @@ func (ed *economicsData) ComputeGasLimit(tx process.TransactionWithFeeHandler) u
...
+	gasRefund := big.NewInt(0).Div(refundValue, gasPriceBig).Uint64()
+
```
Comment (sasurobert) [inline process/economics/economicsData.go:424]: delete empty line

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: process/economics/economicsData.go [hunk 3]
```diff
@@ -389,6 +389,44 @@ func (ed *economicsData) ComputeGasLimit(tx process.TransactionWithFeeHandler) u
...
 
+// ComputeGasUsedAndFeeBasedOnRefundValue will compute gas used value and transaction fee using refund value from a SCR
+func (ed *economicsData) ComputeGasUsedAndFeeBasedOnRefundValue(tx process.TransactionWithFeeHandler, refundValue *big.Int) (uint64, *big.Int) {
+	if refundValue.Cmp(big.NewInt(0)) == 0 {
+		txFee := ed.ComputeTxFee(tx)
+		return tx.GetGasLimit(), txFee
+	}
+
+	gasRefund := ed.computeGasRefund(refundValue, tx.GetGasPrice())
+	gasUsed := tx.GetGasLimit() - gasRefund
```
Comment (SebastianMarian) [inline process/economics/economicsData.go:400]: Use safeSub with log

File: process/economics/economicsData.go [hunk 4]
```diff
@@ -389,6 +389,44 @@ func (ed *economicsData) ComputeGasLimit(tx process.TransactionWithFeeHandler) u
...
+
+	return txFee
+}
+
+func (ed *economicsData) computeGasRefund(refundValue *big.Int, gasPrice uint64) uint64 {
+	gasPriceBig := big.NewInt(0).SetUint64(gasPrice)
+
+	gasRefund := big.NewInt(0).Div(refundValue, gasPriceBig).Uint64()
+
+	gasRefund = uint64(ed.GasPriceModifier() * float64(gasRefund))
```
Comment (SebastianMarian) [inline process/economics/economicsData.go:425]: So we will assume that any refunded value would have been used with gas price modifier applied @sasurobert

Comment (sasurobert) [reply process/economics/economicsData.go:425]: ed.GasPriceModifier return 1.0 when it is disabled

File: core/indexer/common.go [hunk 5]
```diff
@@ -466,14 +471,22 @@ func prepareSerializedDataForATransaction(
...
+		if isSCRForSenderWithRefund(sc, tx) {
 			return true
 		}
 	}
 
 	return false
 }
 
+func isRelayedTx(tx *Transaction) bool {
+	if strings.HasPrefix(string(tx.Data), "relayedTx") && len(tx.SmartContractResults) > 0 {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/common.go:483]: done.

File: core/indexer/factory/indexerFactory.go [hunk 6]
```diff
@@ -159,8 +159,8 @@ func checkDataIndexerParams(arguments *ArgsIndexerFactory) error {
 	if check.IfNil(arguments.EpochStartNotifier) {
 		return core.ErrNilEpochStartNotifier
 	}
-	if arguments.FeeConfig == nil {
-		return core.ErrNilFeeConfig
+	if arguments.TransactionFeeCalculator == nil {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/factory/indexerFactory.go:162]: done.

File: core/indexer/processTransactions.go [hunk 7]
```diff
@@ -115,21 +96,32 @@ func (tdp *txDatabaseProcessor) prepareTransactionsForDatabase(
...
 					continue
 				}
 			}
 
-			if strings.Contains(string(transactions[hash].Data), "relayedTx") {
-				transactions[hash].GasUsed = transactions[hash].GasLimit
+			if isRelayedTx(tx) {
+				tx.GasUsed = tx.GasLimit
+				fee := tdp.txFeeCalculator.ComputeTxFeeBasedOnGasUsed(tx, tx.GasUsed)
+				transactions[hash].Fee = fee.String()
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/processTransactions.go:116]: done.

File: core/indexer/processTransactions.go [hunk 8]
```diff
@@ -115,21 +96,32 @@ func (tdp *txDatabaseProcessor) prepareTransactionsForDatabase(
...
 				continue
 			}
 
-			transactions[hash].Status = transaction.TxStatusFail.String()
+			tx.Status = transaction.TxStatusFail.String()
```
[the remark this answers is not in the document]
Comment (miiu96) [reply core/indexer/processTransactions.go:120]: done.

File: process/economics/economicsData.go (hunk 3 above)
Comment (miiu96) [reply process/economics/economicsData.go:400]: done.

File: process/economics/economicsData.go (hunk 1 above)
Comment (miiu96) [reply process/economics/economicsData.go]: done.

File: process/economics/economicsData.go (hunk 2 above)
Comment (miiu96) [reply process/economics/economicsData.go]: done.

**Review by sasurobert — APPROVED** [review]

File: process/economics/economicsData.go (hunk 4 above)
Comment (SebastianMarian) [reply process/economics/economicsData.go:427]: Yes I know, but I have referred here to all the situations after the flag activation.

**Review by SebastianMarian — APPROVED** [review]

**Review by LucianMincu — APPROVED** [review]
System tests passed

**Pull request merged**